rust/lance-file/src/v2/reader.rs

```diff
@@ -0,0 +1,515 @@
+                let encoding = Self::fetch_encoding(encoding);
+                println!("     Encoding:");
+                println!();
+                let encoding_dbg = format!("{:#?}", encoding);
+                for line in encoding_dbg.lines() {
+                    println!("       {}", line);
+                }
+                println!();
+            }
+        }
+
+        Ok(())
+    }
+
+    fn fetch_encoding(encoding: &pbfile::Encoding) -> pbenc::ArrayEncoding {
+        match &encoding.style {
+            Some(pbfile::encoding::Style::Deferred(_)) => todo!(),
+            Some(pbfile::encoding::Style::Direct(encoding)) => encoding
+                .encoding
+                .as_ref()
+                .unwrap()
+                .to_msg::<pbenc::ArrayEncoding>()
+                .unwrap(),
+            None => panic!(),
+        }
+    }
+
+    fn meta_to_col_infos(meta: &CachedFileMetadata) -> Vec<ColumnInfo> {
+        meta.column_metadatas
+            .iter()
+            .map(|col_meta| {
+                let page_infos = col_meta
+                    .pages
+                    .iter()
+                    .map(|page| {
+                        let num_rows = page.length;
+                        let buffer_offsets = Arc::new(page.buffer_offsets.clone());
+                        let encoding = Self::fetch_encoding(page.encoding.as_ref().unwrap());
+                        Arc::new(PageInfo {
+                            buffer_offsets,
+                            encoding,
+                            num_rows,
+                        })
+                    })
+                    .collect::<Vec<_>>();
+                ColumnInfo {
+                    page_infos,
+                    buffer_offsets: vec![],
+                }
+            })
+            .collect::<Vec<_>>()
+    }
+
+    /// Opens a new file reader without any pre-existing knowledge
+    ///
+    /// This will read the file schema and desired column metadata from the
+    pub async fn try_open(scheduler: FileScheduler, projection: Schema) -> Result<Self> {
+        let file_metadata = Arc::new(Self::read_all_metadata(&scheduler, &projection).await?);
+        let column_infos = Self::meta_to_col_infos(&file_metadata);
+        let num_rows = file_metadata.num_rows;
+        Ok(Self {
+            scheduler: Arc::new(LanceEncodingsIo(scheduler)),
+            column_infos,
+            file_schema: file_metadata.file_schema.clone(),
+            num_rows,
+        })
+    }
+
+    async fn read_range(
+        &self,
+        range: Range<u64>,
+        batch_size: u32,
+    ) -> BoxStream<'static, JoinHandle<Result<RecordBatch>>> {
+        let mut decode_scheduler =
+            DecodeBatchScheduler::new(&self.file_schema, &self.column_infos, &vec![]);
+
+        let (tx, rx) = mpsc::unbounded_channel();
+
+        let scheduler = self.scheduler.clone() as Arc<dyn EncodingsIo>;
+        // FIXME: spawn this, change this method to sync
```

Comment:
What's your philosophy here on what should be sync and what should be async?

Reply:
Decode happens by polling the returned stream.  This gives a rust user "what they expect" when it comes to parallelism / readahead.  If the user does not call `buffered` then there is one decode thread and no "batch parallelism".  If the user calls `buffered` then there are N decode threads / batch parallelism.

Right now that decode won't start until the scheduling has completely finished which is not ideal but not terrible (scheduling is pretty fast).  It'll be more of an issue when there is a chance of backpressure being needed (e.g. decoder falls behind I/O).

If this method is async then it means some kind of "potentially slow" task is happening before the decode thread starts and I can't think of any valid reason for that.

Also, the schedule_range method itself is roughly synchronous.  Ideally it would be entirely synchronous.  Scheduling should never have to pause and wait for I/O (we go through great lengths to avoid this in the list scheduler).  This is because any pause here runs the risk of starving our I/O parallelism.

I haven't yet concluded if I can safely make `schedule_range` a synchronous method though.  My only remaining concern is backpressure.  The scheduler may need to pause if it gets too far ahead of the decoder.  On the other hand, I don't think this is very important because the size of "all scheduled tasks" should be roughly equivalent to the size of the file metadata (e.g. not too large) and so backpressure is more of a concern for the I/O scheduler than it is for the scheduling thread.